### ISSUE TITLE
Local function with [MaybeNullWhen(false)] crashes in nullable analysis

### ISSUE THREAD

**Issue by RikkiGibson** [dotnet/roslyn#41479]
This is a bug in the local-function-attributes feature branch.

Sample:
```cs
#nullable enable

using System.Diagnostics.CodeAnalysis;

class C
{
    void M()
    {
        _ = tryGetValue(out string? s);
        bool tryGetValue<T>([MaybeNullWhen(false)] out T t)
        {
            t = default;
            return false;
        }
    }
}
```

This code throws a NullReferenceException about 'other' being null here:
```
NullableWalker.Join(ref NullableWalker.LocalState self, ref NullableWalker.LocalState other) Line 8247
NullableWalker.VisitLocalFunctionUse(LocalFunctionSymbol symbol, NullableWalker.LocalFunctionState localFunctionState, SyntaxNode syntax, bool isCall) Line 1804
AbstractFlowPass<NullableWalker.LocalState, NullableWalker.LocalFunctionState>.VisitLocalFunctionUse(LocalFunctionSymbol symbol, SyntaxNode syntax, bool isCall) Line 1176
NullableWalker.VisitCall(BoundCall node) Line 3465
```

**Comment by RikkiGibson** [comment]
Note that there's no crash when just using `[MaybeNull]` on the parameter. Not sure yet what the difference is.

It also doesn't trigger when the attribute is just some user-defined attribute not known to NullableWalker.
